From 831a55af755109826e331f96c51ce26de016fa27 Mon Sep 17 00:00:00 2001
From: beck <>
Date: Sun, 6 Nov 2016 10:31:34 +0000
Subject: [PATCH] The upcoming x509 alt chains diff tightens the trust
 requirements for certificates. This (from OpenSSL) ensures that the current
 "default" behaviour remains the same.  We should revisit this later ok jsing@

---
 crypto/x509/x509_trs.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/crypto/x509/x509_trs.c b/crypto/x509/x509_trs.c
index 7a748dd..3f274b5 100644
--- a/crypto/x509/x509_trs.c
+++ b/crypto/x509/x509_trs.c
@@ -116,6 +116,22 @@ X509_check_trust(X509 *x, int id, int flags)
 
 	if (id == -1)
 		return 1;
+	/*
+	 * XXX beck/jsing This enables self signed certs to be trusted for
+	 * an unspecified id/trust flag value (this is NOT the
+	 * X509_TRUST_DEFAULT), which was the longstanding
+	 * openssl behaviour. boringssl does not have this behaviour.
+	 *
+	 * This should be revisited, but changing the default "not default"
+	 * may break things.
+	 */
+	if (id == 0) {
+		int rv;
+		rv = obj_trust(NID_anyExtendedKeyUsage, x, 0);
+		if (rv != X509_TRUST_UNTRUSTED)
+			return rv;
+		return trust_compat(NULL, x, 0);
+	}
 	idx = X509_TRUST_get_by_id(id);
 	if (idx == -1)
 		return default_trust(id, x, flags);
